### ISSUE TITLE
[onedrive] 403 forbidden on some files - try AVOverride

### ISSUE THREAD

**Issue by Darkvater** [rclone/rclone#557]
Using rclone v1.30 with onedrive on an arm. 

I have certain files that I have uploaded using either the browser or the official onedrive windows client that I am unable to download to my machine. 

I have the following outor in verbose mode:

`2016/07/03 11:47:23 Couldn't decode error response: EOF
2016/07/03 11:47:23 DBGT-html.tar.bz2: Failed to copy: failed to open source object: 403 Forbidden:`

I think it could be because onedrive cannot verify the file is not a threat, as when I try to download it through the browser I get a prompt/verification saying:

> Unable to scan DBGT-html.tar.bz2 for viruses
> OneDrive is unable to scan DBGT-html.tar.bz2 for viruses. To download the file, select Continue.
> I understand that downloading this file might harm my computer.
> Continue with download Cancel

The file is of course virus-free. Link to the file: https://1drv.ms/u/s!ApPy0StHU12Kgv4u0O-l4GzT3p3ISw

**Comment by ncw** [comment]
Interesting!

If I try to download the link you sent me then Canel is the only option I get.

I can't do anything about onedrive thinking the file is a threat, so I guess the question is, is there anything in the API which could say "I'm OK with downloading a file which might contain viruses" and allow rclone to download the file.

Onedrive has a `virusSuspicious` error code which you can see here: https://dev.onedrive.com/misc/errors.htm but it doesn't look like it is returning that from the log.

That is the only mention of the word virus in the API docs.

When you try to download the file through the browser and you click OK, can you see what file your browser actually downloads? I'd do this with the [chrome devtools](https://developer.chrome.com/devtools) and the network tab.  Most browsers have something similar now-a-days.  I'm just wondering if there is a `?acknowledgeVirus=true` flag or something.

**Comment by Darkvater** [comment]
Hi ncw, this is the flow when I download the file, I think the magic query parameter is AVOverride=1

1. click on download:
  * Url: https://onedrive.live.com/download.aspx
  * Response Code: 302 (Found)
  * Query String Parameters:
    * cid:[some-cid]
    * resid:[some-cid]!48942
    * canary:ublwZ/2y043MfQibrrX+RlfVzLDRIHNx6JElwS8puSQ=7
    * ithint:.bz2
2. I get forwarded to:
  * Url: https://uagowq-ch3301.files.1drv.com/.../DBGT-html.tar.bz2
  * Response Code: 302 (Found)
  * Response headers: X-Virus-Infected:ScanError
  * Query String Parameters:
    * download:
    * psid:1
3. I get forwarded to:
  * Url: https://skydrive.live.com/redir.aspx
  * Response Code: 200 (OK)
  * Query String Parameters:
    * resid[<some-cid]!48942
    * avres:ScanError
    * averror:DAMAGED_STREAM
    * page:viruswarning
4. Tries to download:
  * Url: https://onedrive.live.com/viruswarning.aspx/Pictures/DBGT-html.tar.bz2
  * Response Code: 403 (Forbidden)
  * Query String Parameters:
    * resid:[some-cid]!48942
    * avres:ScanError
    * averror:DAMAGED_STREAM
    * page:viruswarning
5. I get confirmation, I actually click to download:
  * Url: ttps://uagowq-ch3301.files.1drv.com/<some-link>/DBGT-html.tar.bz2
  * Response Code: 200 (OK)
  * Query String Parameters:
    * download:
    * psid:1
    * **AVOverride:1**

**Comment by paz** [comment]
Was this issue ever resolved?

```
2020/09/06 02:11:29 ERROR : Attempt 3/3 failed with 3 errors and: failed to open source object: 403 Forbidden:
2020/09/06 02:11:29 Failed to sync with 3 errors: last error was: failed to open source object: 403 Forbidden:
```

**Comment by ncw** [comment]
We never implement the avoverride flag - do you want to have a go @paz ?

**Comment by ivandeex** [comment]
@Darkvater @paz Does the problem reproduce with rclone 1.54?

**Comment by paz** [comment]
the problem randomly disappeaared for me on 1.53.3, not sure why. maybe onedrive stopped deciding the file was a virus

**Comment by ivandeex** [comment]
@paz Got you
Closing for now.
Feel free to resubmit

**Comment by voltagex** [comment]
This is reoccurring for me on rclone v1.60.1

A "good" way to reproduce this is:

Download https://www.nirsoft.net/utils/produkey-x64.zip and upload it to OneDrive (https://www.virustotal.com/gui/url/22d3d4791b9a9fc01bd848d418d01dce2cb1560a3d72b2a6631485c0c5abf307)

**Comment by voltagex** [comment]
Hit this again, can this be reopened @ivandeex?

**Comment by ncw** [comment]
@voltagex do you want to implement the `AVOverride=1` flag?

Something like this should do for a proof of concept

```patch
--- a/backend/onedrive/onedrive.go
+++ b/backend/onedrive/onedrive.go
@@ -1966,6 +1966,7 @@ func (o *Object) Open(ctx context.Context, options ...fs.OpenOption) (in io.Read
 	var resp *http.Response
 	opts := o.fs.newOptsCall(o.id, "GET", "/content")
 	opts.Options = options
+	opts.Parameters = url.Values{"AVOverride": {"1"}}
 
 	err = o.fs.pacer.Call(func() (bool, error) {
 		resp, err = o.fs.srv.Call(ctx, &opts)
```

Then it would need to be configured with a flag if that works.

**Comment by voltagex** [comment]
```
voltagex@boxodisks:/tmp/onedrive/test$ ls
ProduKey.chm  ProduKey.exe  produkey-x64.zip  readme.txt
voltagex@boxodisks:/tmp/onedrive/test$ file ProduKey.exe
ProduKey.exe: ERROR: cannot read `ProduKey.exe' (Input/output error)
voltagex@boxodisks:/tmp/onedrive/test$ file produkey-x64.zip
produkey-x64.zip: ERROR: cannot read `produkey-x64.zip' (Input/output error)
voltagex@boxodisks:/tmp/onedrive/test$ rclone version
rclone v1.61.1
- os/version: ubuntu 22.04 (64 bit)
- os/kernel: 5.15.0-71-generic (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.19.4
- go/linking: static
- go/tags: none
```
from rclone:

```
2023/05/03 21:24:06 ERROR : IO error: 403 Forbidden:
2023/05/03 21:24:06 ERROR : IO error: 403 Forbidden:
2023/05/03 21:24:10 ERROR : IO error: 403 Forbidden:
2023/05/03 21:24:11 ERROR : IO error: 403 Forbidden:
```

(there should probably be more info in those log lines!)

With the patch:

```
voltagex@boxodisks:/tmp/onedrive/test$ file produkey-x64.zip
produkey-x64.zip: Zip archive data, at least v2.0 to extract, compression method=deflate
voltagex@boxodisks:/tmp/onedrive/test$ file ProduKey.exe
ProduKey.exe: PE32+ executable (GUI) x86-64, for MS Windows
```

IMO, how this should work:
* A parameter like `--allow-avoverride`

* If we get a 403 back from OneDrive, warn in the log, retry with the AVOverride parameter if the flag is supplied.

**Comment by ncw** [comment]
Please give this a go

[v1.63.0-beta.7008.61d6f538b.fix-557-onedrive-av](https://beta.rclone.org/branch/fix-557-onedrive-av/v1.63.0-beta.7008.61d6f538b.fix-557-onedrive-av/) on branch [fix-557-onedrive-av](https://github.com/rclone/rclone/tree/fix-557-onedrive-av) (uploaded in 15-30 mins)

It will detect blocked files with an error message like this

    ERROR : produkey-x64.zip: Failed to copy: failed to open source object: server reports this file is infected with a virus - use --onedrive-av-override to download anyway: Infected (HackTool:Win32/ProductKey): 403 Forbidden: 

And then you can download the file by using --onedrive-av-override

**Comment by voltagex** [comment]
@ncw confirming this is working for me on Linux, rsyncing from a mounted drive with the new flag.

One minor nit - it'd be nice to still report the files that AVOverride was needed for, but this is no big deal.

Thanks for doing this.
